File: tools/screen_sender.js
const net = require('net');
const screenshot = require('screenshot-desktop');
const sharp = require('sharp');

const ESP_IP = process.argv[2] || '192.168.43.150';
const FPS = parseInt(process.argv[3]) || 15;
const INTERVAL = Math.floor(1000 / FPS);

const W = 240, H = 240, SIZE = W * H * 2;
let sock = null;
let connecting = false;

function connect() {
    if (connecting) return;
    connecting = true;
    sock = new net.Socket();
    sock.setNoDelay(true);
    sock.connect(8888, ESP_IP, () => {
        console.log('Connected to ' + ESP_IP + ':8888');
        connecting = false;
    });
    sock.on('error', () => { sock.destroy(); connecting = false; });
    sock.on('close', () => { sock = null; connecting = false; });
}

async function sendFrame() {
    if (!sock || sock.destroyed) { connect(); return; }
    try {
        const png = await screenshot({ format: 'png' });
        // sharp: native C, much faster than Jimp
        const raw = await sharp(png).resize(W, H).ensureAlpha().raw().toBuffer();

        const frame = Buffer.allocUnsafe(SIZE);
        for (let i = 0, j = 0; i < raw.length; i += 4, j += 2) {
            const v = ((raw[i] & 0xF8) << 8) | ((raw[i+1] & 0xFC) << 3) | (raw[i+2] >> 3);
            frame[j] = v & 0xFF;
            frame[j + 1] = (v >> 8) & 0xFF;
        }
        sock.write(frame);
    } catch (e) { /* skip */ }
}

setInterval(sendFrame, INTERVAL);
connect();
console.log('Streaming ' + FPS + 'fps -> ' + ESP_IP + ':8888');

File: tools/screen_sender.py
"""
Screen stream sender for ESP32-S3 MP4.
Captures desktop, resizes to 240x240, sends raw RGB565 over TCP.

Usage: python screen_sender.py <ESP32_IP> [fps]
  e.g.  python screen_sender.py 192.168.43.150 10

Requirements: pip install pillow
"""
import socket
import sys
import time
from PIL import Image, ImageGrab

PORT = 8888
W, H = 240, 240

sock = None


def connect(host):
    global sock
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        s.settimeout(3)
        s.connect((host, PORT))
        sock = s
        print(f"Connected to {host}:{PORT}")
        return True
    except Exception as e:
        print(f"Connect failed: {e}, retrying...")
        sock = None
        return False


def send_frame():
    global sock
    if sock is None:
        return
    try:
        img = ImageGrab.grab()
        img = img.resize((W, H), Image.LANCZOS).convert("RGB")
        pixels = img.load()
        buf = bytearray(W * H * 2)
        idx = 0
        for y in range(H):
            for x in range(W):
                r, g, b = pixels[x, y]
                val = ((r & 0xF8) << 8) | ((g & 0xFC) << 3) | (b >> 3)
                buf[idx] = val & 0xFF
                buf[idx + 1] = (val >> 8) & 0xFF
                idx += 2
        sock.sendall(bytes(buf))
    except Exception:
        sock = None


def main(host, fps):
    print(f"Streaming {fps}fps -> {host}:{PORT}")
    interval = 1.0 / fps

    while True:
        if sock is None:
            connect(host)
            time.sleep(1)
            continue

        t0 = time.time()
        send_frame()
        elapsed = time.time() - t0
        if elapsed < interval:
            time.sleep(interval - elapsed)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Usage: python {sys.argv[0]} <ESP32_IP> [fps]")
        sys.exit(1)
    main(sys.argv[1], int(sys.argv[2]) if len(sys.argv) > 2 else 10)
